### ISSUE TITLE
S3 - Block Public S3 Object ACLs - Lambda function recursive trigger

### ISSUE THREAD

**Issue by karthikeayan** [cloud-custodian/cloud-custodian#3238]
This policy have two event source, 1. CreateBucket, 2. PutBucketPolicy.

So below is the event flow when someone creating a S3 Bucket.

1. User creates a S3 Bucket
2. Cloudwatch event triggered for the new bucket creation
3. Lambda function triggered
4. Lambda function adds a policy to the bucket(PutBucketPolicy)
5. Cloudwatch event triggered for the PutBucketPolicy
6. Lambda function triggered
.
.
.

So, steps 4, 5, 6 executes recursively and I can see around 4000 lambda invocations for any single s3 bucket creation. Any one facing the same issue?

**Comment by thisisshi** [comment]
you should be authoring your policy in such a way to not filter in buckets that have the correct configuration to prevent this type of recursive lambda execution, for example

```yaml
policies:
  - name: s3-no-public
    resource: s3
    mode:
      type: cloudtrail
      events:
        - CreateBucket
        - PutBucketPolicy
    filters:
        - type: cross-account
    actions:
        - type: remove-statements
          statement_ids: matched
```

In this case, the lambda function would fire at most 2 times for a single bucket/policy change, as after the first execution, the bucket policy will not match the filters that were passed in

**Comment by kapilt** [comment]
What @thisisshi said.. policies should converge, and should only be filtering to things of interest, events subscriptions are not filters they are sources, ideally r/w event policies that are modifying the resource do so in such a way that it does not continue to pass filters, else yeah you get a loop, **as expected.**

**Comment by karthikeayan** [comment]
Am doing the below, but still somehow am getting the loop.

```
    mode:
      type: cloudtrail
      events:
        - CreateBucket
        - source: 's3.amazonaws.com'
          event: DeleteBucketPolicy
          ids: "requestParameters.bucketName"
        - source: 's3.amazonaws.com'
          event: PutBucketPolicy
          ids: "requestParameters.bucketName"
      role: arn:aws:iam::<my_account>:role/CloudCustodianS3ObjectPublicDenyACL
      timeout: 200
    description: |
      Appends a bucket policy statement to an s3 bucket when it detects
      a policy change to the bucket or a new bucket is created which
      will deny anyone from setting s3 objects in the bucket to public-read,
      public-read-write, or any authenticated AWS user.
    filters:
      - type: missing-statement
        statement_ids:
          - DenyS3PublicObjectACL
    actions:
      - type: set-statements
        statements:
          - Sid: "DenyS3PublicObjectACL"
            Effect: "Deny"
```

**Comment by karthikeayan** [comment]
The above policy which I created seems working fine. The issue turns out to be, I haven't given the getBucketPolicy and getBucketAcl permission to the lambda function. Once I given the required permission the lambda function triggers twice and the second trigger doesn't filter anything, so the loop ends.

**Comment by rikkuness** [comment]
Does this mean the policy example on https://cloudcustodian.io/docs/aws/examples/s3denypublicobjectacls.html is missing some configuration to prevent a feedback loop?

**Comment by kapilt** [comment]
no, because the filter checks for the condition before applying the action that makes the resource no longer match on a subsequent execution.

**Comment by rikkuness** [comment]
Yeah I think I get you, the example in the docs for `s3-deny-public-object-acl-realtime` doesn't have any filter defined though? Or am I missing something?

I got there in the end anyway, but when I verbatim copied the docs example I got the looping until I realised I needed to stop it picking up it's own changes. Wasn't sure if I was deploying it in a weird way to cause that or the docs just needed a note to point out it might happen.

Thanks for the fast response by the way. :)
